### DISCUSSION TITLE
In what commit/verison were braces allowed? Thank You!

### DISCUSSION THREAD

**Discussion by rickswe** [fish-shell/fish-shell#10333] [Q&A] [answered]
I accidentally discovered this by accident today. This was not possible in the past.
I did look at the [change log](https://fishshell.com/docs/current/relnotes.html). I think it could have been in commit #6415 ?

```sh
mkdir -p ~/Downloads/testfish/{one,two/three}
```

Thanks

**Comment by faho** [comment]
[Brace expansion](https://fishshell.com/docs/current/language.html#brace-expansion) has basically always been in fish.

This has nothing to do with 6415, which is about command grouping.

**Reply by faho** [reply]
Yup, it's been [included in the initial git revision](https://github.com/fish-shell/fish-shell/commit/149594f974350bb364a76c73b91b1d5ffddaa1fa#diff-4e9f03b0fcb26abe26e8d83441df7ea667ec7e16089a316713421ec74247dce7R850-R853), which shows changelogs mentioning them going back to 2005. (at that time it was called "bracket expansion")

**Reply by rickswe** [reply]
Is there anything in fish having to do with braces `{}` that works in bash but does not
work in fish? Cause I remember I had to use a bash shell for something,
I don't think it was command grouping.

**Reply by faho** [reply]
Fish exclusively uses `{}` for `echo {/usr,}/bin` style brace-expansion.

Bash *also* uses them for command grouping and for `${foo}` variable expansions (including `${foo/bar/baz}` and `${foo%bar}` and `${foo[@]}`. See http://mywiki.wooledge.org/BashSheet (and search for `{`).

So yes, these symbols are less overloaded.

Fish has alternatives for all of these - the string manipulations would be done via [string](https://fishshell.com/docs/current/cmds/string.html), expanding arrays is just the default - `$foo` expands to all elements of $foo, so it's equivalent to `"${foo[@]}"`. Command grouping is `begin` and `end`.

**Reply by rickswe** [reply]
@faho  I think it was grouping commands that was not working or is still not working?

```sh
 ls foo || { echo "HI" ;}
ls: cannot access 'foo': No such file or directory
fish: Unknown command: '{ echo HI ;}'
fish: '{ ... }' is not supported for grouping commands. Please use 'begin; ...; end'
ls foo || { echo "HI" ;}
          ^~~~~~~~~~~~~^
```

**Reply by faho** [reply, marked as answer]
Fish uses `begin` and `end` for command grouping, not braces:

```fish
ls foo || begin; echo "HI"; end
```

That is also what the error message tells you.

This is on purpose because it means that the `{}` symbols are less overloaded. Fish does not try to be bash-compatible and it doesn't shoot for code golf. It is fine if something is a word instead of a symbol, especially when that symbol is already used for something else.
